File: app/src/main/java/net/mbonnin/arcanetracker/adapter/Controller.java
package net.mbonnin.arcanetracker.adapter;

import android.content.Context;
import android.os.Handler;
import android.text.TextUtils;

import com.annimon.stream.Collector;
import com.annimon.stream.Stream;
import com.annimon.stream.function.BiConsumer;
import com.annimon.stream.function.Function;
import com.annimon.stream.function.Supplier;

import net.mbonnin.arcanetracker.ArcaneTrackerApplication;
import net.mbonnin.arcanetracker.CardUtil;
import net.mbonnin.arcanetracker.Deck;
import net.mbonnin.arcanetracker.R;
import net.mbonnin.arcanetracker.Utils;
import net.mbonnin.arcanetracker.parser.Entity;
import net.mbonnin.arcanetracker.parser.EntityList;
import net.mbonnin.arcanetracker.parser.Game;
import net.mbonnin.arcanetracker.parser.GameLogic;
import net.mbonnin.hsmodel.PlayerClass;
import net.mbonnin.hsmodel.Rarity;
import net.mbonnin.hsmodel.Type;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

public class Controller implements GameLogic.Listener {
    private static Controller sController;

    private final ItemAdapter mPlayerAdapter;
    private final Handler mHandler;
    private final ItemAdapter mOpponentAdapter;
    protected HashMap<String, Integer> mPlayerCardMap;
    private Game mGame;
    private String mPlayerId;
    private String mOpponentId;

    public Controller() {
        mPlayerAdapter = new ItemAdapter();
        mOpponentAdapter = new ItemAdapter();
        GameLogic.get().addListener(this);
        mHandler = new Handler();
    }

    public ItemAdapter getPlayerAdapter() {
        return mPlayerAdapter;
    }

    public ItemAdapter getOpponentAdapter() {
        return mOpponentAdapter;
    }

    public void setPlayerDeck(HashMap<String, Integer> cardMap) {
        mPlayerCardMap = cardMap;
        update();
    }

    private Runnable mUpdateRunnable = this::update;

    private ArrayList<Object> entityListToItemList(EntityList entityList, Function<Entity, Boolean> increasesCount) {
        /*
         * remove and count the really unknown cards
         */
        int unknownCards = 0;
        Iterator<Entity> it = entityList.iterator();
        while (it.hasNext()) {
            Entity entity = it.next();
            if (!entity.extra.tmpIsGift
                    && entity.extra.tmpCard == CardUtil.UNKNOWN) {
                it.remove();
                unknownCards++;
            }
        }

        /*
         * Sort and merge
         */
        Comparator<Entity> comparator = (a, b) -> {
            int acost = a.extra.tmpCard.cost == null ? 0 : a.extra.tmpCard.cost;
            int bcost = b.extra.tmpCard.cost == null ? 0 : b.extra.tmpCard.cost;
            if (acost < 0) {
                acost = Integer.MAX_VALUE;
            }
            if (bcost < 0) {
                bcost = Integer.MAX_VALUE;
            }

            int ret = acost - bcost;

            if (ret != 0) {
                return ret;
            }

            ret = a.extra.tmpCard.name.compareTo(b.extra.tmpCard.name);
            if (ret != 0) {
                return ret;
            }

            int aGift = a.extra.tmpIsGift ? 1 : 0;
            int bGift = b.extra.tmpIsGift ? 1 : 0;

            return bGift - aGift;
        };
        Collections.sort(entityList, comparator);

        Collector<Entity, ArrayList<DeckEntryItem>, ArrayList<DeckEntryItem>> collector = new Collector<Entity, ArrayList<DeckEntryItem>, ArrayList<DeckEntryItem>>() {

            @Override
            public Supplier<ArrayList<DeckEntryItem>> supplier() {
                return ArrayList::new;
            }

            @Override
            public BiConsumer<ArrayList<DeckEntryItem>, Entity> accumulator() {
                return (list, entity) -> {
                    DeckEntryItem deckEntry;
                    if (list.size() == 0 || entity.extra.tmpIsGift) {
                        deckEntry = new DeckEntryItem();
                        list.add(deckEntry);
                    } else {
                        deckEntry = list.get(list.size() - 1);
                        if (entity.extra.tmpCard != deckEntry.card) {
                            deckEntry = new DeckEntryItem();
                            list.add(deckEntry);
                        }
                    }
                    deckEntry.entityList.add(entity);
                    if (increasesCount.apply(entity)) {
                        deckEntry.count++;
                    }
                    deckEntry.card = entity.extra.tmpCard;
                    deckEntry.gift = entity.extra.tmpIsGift;
                };
            }

            @Override
            public Function<ArrayList<DeckEntryItem>, ArrayList<DeckEntryItem>> finisher() {
                return list -> list;
            }
        };

        ArrayList<DeckEntryItem> deckEntryItemList = Stream.of(entityList).collect(collector);

        /*
         * sort the entity list
         */
        for (DeckEntryItem deckEntryItem : deckEntryItemList) {
            Collections.sort(deckEntryItem.entityList, (a, b) -> a.extra.drawTurn - b.extra.drawTurn);
        }

        ArrayList<Object> itemList = new ArrayList<>();
        itemList.addAll(deckEntryItemList);
        if (unknownCards > 0) {
            itemList.add(ArcaneTrackerApplication.getContext().getString(R.string.unknown_cards, unknownCards));
        }

        return itemList;
    }

    /*
     * this attempts to map the knowledge that we have of mDeck to the unknown entities
     * this assumes that an original card is either known or still in deck
     * this sets tmpCard
     *
     * what needs to be handled is hemet, maybe others ?
     */
    private void assignCardsFromDeck() {
        EntityList originalDeckEntityList = mGame.getEntityList(entity -> mPlayerId.equals(entity.extra.originalController));
        ArrayList<String> cardIdsFromDeck = new ArrayList<>();

        /*
         * build a list of all the ids in mDeck
         */
        for (Map.Entry<String, Integer> entry : mPlayerCardMap.entrySet()) {
            for (int i = 0; i < entry.getValue(); i++) {
                cardIdsFromDeck.add(entry.getKey());
            }
        }

        /*
         * remove the ones that have been revealed already
         */
        for (Entity entity : originalDeckEntityList) {
            if (!TextUtils.isEmpty(entity.CardID)) {
                Iterator<String> it = cardIdsFromDeck.iterator();
                while (it.hasNext()) {
                    String next = it.next();
                    if (next.equals(entity.CardID)) {
                        it.remove();
                        break;
                    }
                }
            }
        }

        /*
         * assign a tmpCard to the cards we still don't know
         */
        int i = 0;
        for (Entity entity : originalDeckEntityList) {
            if (entity.card == null) {
                if (i < cardIdsFromDeck.size()) {
                    entity.extra.tmpCard = CardUtil.INSTANCE.getCard(cardIdsFromDeck.get(i));
                    i++;
                }
            }
        }
    }

    private EntityList getEntityListInZone(String playerId, String zone) {
        return mGame.getEntityList(entity -> playerId.equals(entity.tags.get(Entity.KEY_CONTROLLER))
                && zone.equals(entity.tags.get(Entity.KEY_ZONE)));
    }

    private static int compareNullSafe(String a, String b) {
        if (a == null) {
            return b == null ? 0 : 1;
        } else {
            return b == null ? -1 : a.compareTo(b);
        }
    }

    private ArrayList getHand() {
        ArrayList<Object> list = new ArrayList<>();
        Context context = ArcaneTrackerApplication.getContext();

        EntityList entities = getEntityListInZone(mOpponentId, Entity.ZONE_HAND);

        Collections.sort(entities, (a, b) -> compareNullSafe(a.tags.get(Entity.KEY_ZONE_POSITION), b.tags.get(Entity.KEY_ZONE_POSITION)));

        list.add(new HeaderItem(context.getString(R.string.hand) + " (" + entities.size() + ")"));
        for (Entity entity : entities) {
            DeckEntryItem deckEntry = new DeckEntryItem();
            if (TextUtils.isEmpty(entity.CardID) || entity.extra.hide) {
                deckEntry.card = CardUtil.INSTANCE.unknown();
                StringBuilder builder = new StringBuilder();
                builder.append("#").append(GameLogic.gameTurnToHumanTurn(entity.extra.drawTurn));
                if (entity.extra.mulliganed) {
                    builder.append(" (M)");
                }
                deckEntry.card.name = builder.toString();
            } else {
                deckEntry.card = entity.card;
            }
            deckEntry.gift = !entity.extra.hide && entity.extra.tmpIsGift;
            deckEntry.count = 1;
            Entity clone = entity.clone();
            if (entity.extra.hide) {
                clone.extra.createdBy = null;
            }
            deckEntry.entityList.add(clone);
            list.add(deckEntry);
        }

        return list;
    }


    private void update() {
        if (mGame == null) {
            mPlayerAdapter.setList(getCardMapList(mPlayerCardMap != null ? mPlayerCardMap : new HashMap<>()));
            mOpponentAdapter.setList(getCardMapList(new HashMap<>()));
        } else {
            /*
             * all the code below uses tmpCard and tmpIsGift so that it can change them without messing up the internal game state
             */
            EntityList allEntities = mGame.getEntityList(entity -> true);
            Stream.of(allEntities).forEach(entity -> {
                entity.extra.tmpIsGift = !TextUtils.isEmpty(entity.extra.createdBy);
                if (entity.card != null) {
                    entity.extra.tmpCard = entity.card;
                } else {
                    entity.extra.tmpCard = CardUtil.UNKNOWN;
                }
            });

            updatePlayer();
            updateOpponent();
        }
    }

    private void updateOpponent() {
        ArrayList<Object> list = new ArrayList<>();

        Collection<?> secrets = getSecrets();
        if (secrets.size() > 0) {
            list.add(new HeaderItem(Utils.INSTANCE.getString(R.string.secrets)));
            list.addAll(secrets);
        }
        list.addAll(getHand());

        list.add(new HeaderItem(Utils.INSTANCE.getString(R.string.allCards)));

        // trying a definition that's a bit different from the player definition here
        EntityList allEntities = mGame.getEntityList(e -> mOpponentId.equals(e.tags.get(Entity.KEY_CONTROLLER))
                && !Entity.ZONE_SETASIDE.equals(e.tags.get(Entity.KEY_ZONE))
                && !Type.ENCHANTMENT.equals(e.tags.get(Entity.KEY_CARDTYPE))
                && !Type.HERO.equals(e.tags.get(Entity.KEY_CARDTYPE))
                && !Type.HERO_POWER.equals(e.tags.get(Entity.KEY_CARDTYPE)));

        list.addAll(entityListToItemList(allEntities, e -> true));

        mOpponentAdapter.setList(list);
    }

    private void updatePlayer() {
        ArrayList<Object> list = new ArrayList<>();

        list.add(new HeaderItem(Utils.INSTANCE.getString(R.string.deck)));

        if (mPlayerCardMap != null) {
            assignCardsFromDeck();
        }

        EntityList entityList = mGame.getEntityList(entity -> mPlayerId.equals(entity.extra.originalController));
        /*
         * Add all the gifts
         * XXX it's not enough to filter on !TextUtils.isEmpty(createdBy)
         * because then we get all enchantments
         * if a gift is in the graveyard, it won't be shown but I guess that's ok
         */
        entityList.addAll(mGame.getEntityList(entity -> {
            return Entity.ZONE_DECK.equals(entity.tags.get(Entity.KEY_ZONE))
                    && !mPlayerId.equals(entity.extra.originalController)
                    && mPlayerId.equals(entity.tags.get(Entity.KEY_CONTROLLER));
        }));

        list.addAll(entityListToItemList(entityList, entity -> Entity.ZONE_DECK.equals(entity.tags.get(Entity.KEY_ZONE))));

        mPlayerAdapter.setList(list);
    }

    private Collection<?> getSecrets() {
        ArrayList<Object> list = new ArrayList<>();

        if (false) {
            Stream.of(getEntityListInZone(mOpponentId, Entity.ZONE_HAND)).forEach(e -> {
                e.tags.put(Entity.KEY_CLASS, PlayerClass.MAGE);
                e.tags.put(Entity.KEY_ZONE, Entity.ZONE_SECRET);
                e.extra.drawTurn = 18;
                e.extra.playTurn = 23;
            });
        }
        EntityList entities = getEntityListInZone(mOpponentId, Entity.ZONE_SECRET)
                .filter(e -> !Rarity.LEGENDARY.equals(e.tags.get(Entity.KEY_RARITY))); // remove quests

        Collections.sort(entities, (a, b) -> compareNullSafe(a.tags.get(Entity.KEY_ZONE_POSITION), b.tags.get(Entity.KEY_ZONE_POSITION)));

        for (Entity entity : entities) {
            DeckEntryItem deckEntry = new DeckEntryItem();
            if (TextUtils.isEmpty(entity.CardID)) {
                String clazz = entity.tags.get(Entity.KEY_CLASS);

                if (clazz != null) {
                    deckEntry.card = CardUtil.INSTANCE.secret(clazz);
                } else {
                    deckEntry.card = CardUtil.INSTANCE.secret("MAGE");
                }
            } else {
                deckEntry.card = entity.card;
            }
            deckEntry.gift = entity.extra.tmpIsGift;
            deckEntry.count = 1;

            Entity clone = entity.clone();
            clone.card = deckEntry.card;
            deckEntry.entityList.add(clone);
            list.add(deckEntry);
        }

        return list;
    }

    private static ArrayList<Object> getCardMapList(HashMap<String, Integer> cardMap) {
        ArrayList<Object> list = new ArrayList<>();
        int unknown = Deck.MAX_CARDS;

        for (Map.Entry<String, Integer> entry : cardMap.entrySet()) {
            DeckEntryItem deckEntry = new DeckEntryItem();
            deckEntry.card = CardUtil.INSTANCE.getCard(entry.getKey());
            deckEntry.count = entry.getValue();
            list.add(deckEntry);
            unknown -= deckEntry.count;
        }

        Collections.sort(list, (a, b) -> {
            DeckEntryItem da = (DeckEntryItem) a;
            DeckEntryItem db = (DeckEntryItem) b;

            int acost = da.card.cost == null ? 0 : da.card.cost;
            int bcost = db.card.cost == null ? 0 : db.card.cost;
            if (acost < 0) {
                acost = Integer.MAX_VALUE;
            }
            if (bcost < 0) {
                bcost = Integer.MAX_VALUE;
            }
            return acost - bcost;
        });

        if (unknown > 0) {
            list.add(ArcaneTrackerApplication.getContext().getString(R.string.unknown_cards, unknown));
        }
        return list;
    }


    public void resetGame() {
        mGame = null;
        update();
    }

    @Override
    public void gameStarted(Game game) {
        mGame = game;
        mPlayerId = mGame.player.entity.PlayerID;
        mOpponentId = mGame.opponent.entity.PlayerID;
        update();
    }

    @Override
    public void gameOver() {

    }

    @Override
    public void somethingChanged() {
        /*
         * we gate the notification so as not to flood the listeners
         */
        mHandler.removeCallbacks(mUpdateRunnable);
        mHandler.postDelayed(mUpdateRunnable, 200);
    }


    public static Controller get() {
        if (sController == null) {
            sController = new Controller();
        }

        return sController;
    }

    public static void resetAll() {
        get().resetGame();
    }
}

File: app/src/main/java/net/mbonnin/arcanetracker/adapter/Controller.kt
package net.mbonnin.arcanetracker.adapter


import android.os.Handler
import android.text.TextUtils
import net.mbonnin.arcanetracker.*
import net.mbonnin.arcanetracker.parser.Entity
import net.mbonnin.arcanetracker.parser.Game
import net.mbonnin.arcanetracker.parser.GameLogic
import net.mbonnin.hsmodel.enum.PlayerClass
import net.mbonnin.hsmodel.enum.Rarity
import net.mbonnin.hsmodel.enum.Type
import timber.log.Timber
import java.util.*

class Controller : GameLogic.Listener {

    val playerAdapter: ItemAdapter
    val opponentAdapter: ItemAdapter

    private val mHandler: Handler
    protected var mPlayerCardMap: HashMap<String, Int>? = null
    private var mGame: Game? = null
    private var mPlayerId: String? = null
    private var mOpponentId: String? = null

    private val mUpdateRunnable = Runnable { this.update() }

    private val hand: ArrayList<*>
        get() {
            val list = ArrayList<Any>()
            val context = ArcaneTrackerApplication.context

            val entities = getEntityListInZone(mOpponentId, Entity.ZONE_HAND)

            Collections.sort(entities) { a, b -> compareNullSafe(a.tags[Entity.KEY_ZONE_POSITION], b.tags[Entity.KEY_ZONE_POSITION]) }

            list.add(HeaderItem(context.getString(R.string.hand) + " (" + entities.size + ")"))
            for (entity in entities) {
                val card = if (TextUtils.isEmpty(entity.CardID) || entity.extra.hide) {
                    val builder = StringBuilder()
                    builder.append("#").append(GameLogic.gameTurnToHumanTurn(entity.extra.drawTurn))
                    if (entity.extra.mulliganed) {
                        builder.append(" (M)")
                    }
                    CardUtil.unknown(builder.toString())
                } else {
                    entity.card
                }

                val clone = entity.clone()
                if (entity.extra.hide) {
                    clone.extra.createdBy = null
                }

                val deckEntry = DeckEntryItem(
                        card = card!!,
                        gift = !entity.extra.hide && !entity.extra.createdBy.isNullOrEmpty(),
                        count = 1,
                        entityList = listOf(clone)
                )

                list.add(deckEntry)
            }

            return list
        }

    fun getSecrets(): List<DeckEntryItem> {
        val list = ArrayList<DeckEntryItem>()

        val entities = getEntityListInZone(mOpponentId, Entity.ZONE_SECRET)
                .filter { e -> Rarity.LEGENDARY != e.tags[Entity.KEY_RARITY] }
                .sortedBy { it.tags[Entity.KEY_ZONE_POSITION] }

        for (entity in entities) {
            val card = if (TextUtils.isEmpty(entity.CardID)) {
                val clazz = entity.tags[Entity.KEY_CLASS]

                if (clazz != null) {
                    CardUtil.secret(clazz)
                } else {
                    CardUtil.secret("MAGE")
                }
            } else {
                entity.card
            }

            val clone = entity.clone()
            clone.card = card
            val deckEntry = DeckEntryItem(
                    card = card!!,
                    gift = !entity.extra.createdBy.isNullOrEmpty(),
                    count = 1,
                    entityList = listOf(clone)
            )
            list.add(deckEntry)
        }

        return list
    }

    fun getTestSecrets(): List<DeckEntryItem> {
        val list = ArrayList<DeckEntryItem>()

        val entity = Entity()
        entity.tags[Entity.KEY_ZONE] = Entity.ZONE_SECRET
        entity.tags[Entity.KEY_CLASS] = PlayerClass.MAGE
        val deckEntry = DeckEntryItem(
                card = CardUtil.secret("MAGE"),
                gift = false,
                count = 1,
                entityList = listOf(entity)
        )

        list.add(deckEntry)

        return list
    }


    init {
        opponentAdapter = ItemAdapter()
        playerAdapter = ItemAdapter()

        GameLogic.get().addListener(this)
        mHandler = Handler()
    }

    fun setPlayerCardMap(cardMap: HashMap<String, Int>) {
        mPlayerCardMap = cardMap
        update()
    }

    private fun getEntityListInZone(playerId: String?, zone: String): List<Entity> {
        return mGame!!.getEntityList { entity -> playerId == entity.tags[Entity.KEY_CONTROLLER] && zone == entity.tags[Entity.KEY_ZONE] }
    }


    private fun update() {
        if (mGame == null) {
            if (TestSwitch.SECRET_LAYOUT) {
                playerAdapter.setList(getTestSecrets() as ArrayList<Any>)
            } else {
                playerAdapter.setList(getCardMapList(if (mPlayerCardMap != null) mPlayerCardMap!! else HashMap<String, Int>()))
            }

            val list = getCardMapList(HashMap())
            opponentAdapter.setList(list)
        } else {

            playerAdapter.setList(getPlayerList(mPlayerCardMap ?: emptyMap()))
            opponentAdapter.setList(getOpponentList())
        }
    }

    private fun getOpponentList(): List<Any> {
        val list = ArrayList<Any>()

        val secrets = getSecrets()
        if (secrets.size > 0) {
            list.add(HeaderItem(Utils.getString(R.string.secrets)))
            list.addAll(secrets)
        }
        list.addAll(hand)

        list.add(HeaderItem(Utils.getString(R.string.allCards)))

        // trying a definition that's a bit different from the player definition here
        val allEntities = mGame!!.getEntityList { e ->
            (mOpponentId == e.tags[Entity.KEY_CONTROLLER]
                    && Entity.ZONE_SETASIDE != e.tags[Entity.KEY_ZONE]
                    && Type.ENCHANTMENT != e.tags[Entity.KEY_CARDTYPE]
                    && Type.HERO != e.tags[Entity.KEY_CARDTYPE]
                    && Type.HERO_POWER != e.tags[Entity.KEY_CARDTYPE]
                    && "PLAYER" != e.tags[Entity.KEY_CARDTYPE])
        }

        val sanitizedEntities = sanitizeEntities(allEntities)

        val intermediateList = mutableListOf<Intermediate>()
        var unknownCards = 0
        sanitizedEntities.forEach {
            if (it.CardID == null && it.extra.createdBy.isNullOrEmpty()) {
                unknownCards++
            } else {
                intermediateList.add(Intermediate(it.CardID, it))
            }
        }

        list.addAll(intermediateToDeckEntryList(intermediateList, { true }))
        if (unknownCards > 0){
            list.add(ArcaneTrackerApplication.context.getString(R.string.unknown_cards, unknownCards))
        }
        return list
    }

    private fun sanitizeEntities(entityList: List<Entity>): List<Entity> {
        return entityList.map {
            if (it.extra.hide) {
                //if the card is hidden, we don't want to disclose when it was drawn
                val clone = it.clone()
                clone.extra.drawTurn = -1
                clone
            } else {
                it
            }
        }
    }

    class Intermediate(val cardId: String?, val entity: Entity)

    private fun getPlayerList(cardMap: Map<String, Int>): List<Any> {
        val list = mutableListOf<Any>()

        list.add(HeaderItem(Utils.getString(R.string.deck)))

        val knownIdList = ArrayList<String>()

        val intermediateList = mutableListOf<Intermediate>()
        /*
         * build a list of all the ids that we know from the deck or from whizbang
         */
        for ((key, value) in cardMap) {
            for (i in 0 until value) {
                knownIdList.add(key)
            }
        }

        val originalDeckEntityList = mGame!!.getEntityList { entity -> mPlayerId == entity.extra.originalController }

        /*
         * remove the ones that have been revealed already
         */
        val revealedEntityList = originalDeckEntityList.filter { !it.CardID.isNullOrBlank() }
        for (entity in revealedEntityList) {
            val it = knownIdList.iterator()
            while (it.hasNext()) {
                val next = it.next()
                if (next == entity.CardID) {
                    it.remove()
                    break
                }
            }
        }

        /*
         * add the revealed cards
         */
        intermediateList.addAll(revealedEntityList.map { Intermediate(it.CardID, it) })

        /*
         * add the known cards from the deck, assume they are still inside the deck
         */
        intermediateList.addAll(knownIdList.map { Intermediate(it, Entity.UNKNOWN_ENTITY) })

        /*
         * Add all the gifts
         * XXX it's not enough to filter on !TextUtils.isEmpty(createdBy)
         * because then we get all enchantments
         * if a gift is in the graveyard, it won't be shown but I guess that's ok
         */
        val giftList = mGame!!.getEntityList { entity ->
            Entity.ZONE_DECK == entity.tags[Entity.KEY_ZONE]
                    && mPlayerId != entity.extra.originalController
                    && mPlayerId == entity.tags[Entity.KEY_CONTROLLER]
        }

        intermediateList.addAll(giftList.map { Intermediate(it.CardID, it) })

        list.addAll(intermediateToDeckEntryList(intermediateList, { it == Entity.UNKNOWN_ENTITY || it.tags[Entity.KEY_ZONE] == Entity.ZONE_DECK }))

        /*
         * and the unknown if any
         */
        val unknownCards = originalDeckEntityList.size - revealedEntityList.size - knownIdList.size
        if (unknownCards > 0) {
            list.add(ArcaneTrackerApplication.context.getString(R.string.unknown_cards, unknownCards))
        }
        if (unknownCards < 0) {
            Timber.e("too many known card ids: $unknownCards")
        }

        return list
    }

    data class GroupingKey(val cardId: String?, val gift: Boolean)

    private fun intermediateToDeckEntryList(intermediateList: List<Intermediate>, increasesCount: (Entity) -> Boolean): List<DeckEntryItem> {
        val map = intermediateList.groupBy({ GroupingKey(it.cardId, !it.entity.extra.createdBy.isNullOrEmpty()) }, { it.entity })

        val deckEntryList = map.map {
            val cardId = it.key.cardId
            val card = if (cardId == null) {
                CardUtil.UNKNOWN
            } else {
                CardUtil.getCard(cardId)
            }
            val entityList = it.value

            val count = entityList
                    .map { if (increasesCount(it)) 1 else 0 }
                    .sum()

            DeckEntryItem(card,
                    gift = it.key.gift,
                    count = count,
                    entityList = entityList)
        }

        return deckEntryList.sortedBy {
            val costString = it.card.cost?.toString() ?: ""
            val giftString = if(it.gift)  "b" else "a"

            "$costString${it.card.name}$giftString"
        }
    }


    override fun gameStarted(game: Game) {
        mGame = game
        mPlayerId = mGame!!.player!!.entity!!.PlayerID
        mOpponentId = mGame!!.opponent!!.entity!!.PlayerID
        update()
    }

    override fun gameOver() {

    }

    override fun somethingChanged() {
        /*
         * we gate the notification so as not to flood the listeners
         */
        mHandler.removeCallbacks(mUpdateRunnable)
        mHandler.postDelayed(mUpdateRunnable, 200)
    }

    companion object {
        private var sController: Controller? = null

        private fun compareNullSafe(a: String?, b: String?): Int {
            return if (a == null) {
                if (b == null) 0 else 1
            } else {
                if (b == null) -1 else a.compareTo(b)
            }
        }

        fun getCardMapList(cardMap: Map<String, Int>): ArrayList<Any> {
            val list = ArrayList<Any>()
            var unknown = Deck.MAX_CARDS

            for ((key, value) in cardMap) {
                val deckEntry = DeckEntryItem(
                        card = CardUtil.getCard(key),
                        count = value,
                        entityList = emptyList())

                list.add(deckEntry)
                unknown -= deckEntry.count
            }

            Collections.sort(list) { a, b ->
                val da = a as DeckEntryItem
                val db = b as DeckEntryItem

                Utils.compareNullSafe(da.card.cost, db.card.cost)
            }

            if (unknown > 0) {
                list.add(ArcaneTrackerApplication.context.getString(R.string.unknown_cards, unknown))
            }
            return list
        }


        fun get(): Controller {
            if (sController == null) {
                sController = Controller()
            }

            return sController!!
        }
    }
}
